### DISCUSSION TITLE
Bash problem

### DISCUSSION THREAD

**Discussion by tttienthinh** [Taitava/obsidian-shellcommands#280] [Debugging and testing]
I would like to use Pandoc as follow :  
```pandoc {{file_path:absolute}} -o {{folder_path:absolute}}/{{title}}.pdf```  

I am on Ubuntu.  
When I use Environment Linux (Use system default (sh)) : It returns error EscapeValue() : Unrecognised shell: sh
When I use Environment Linux (BASH) : It returns error pdflatex not found

**Comment by Taitava** [comment]
> When I use Environment Linux (Use system default (sh)) : It returns error EscapeValue() : Unrecognised shell: sh

Thank you for letting me know of this! 🙂 It's a bug and I created a new  issue for it: #281 . Edit 2022-11-26: It's now fixed.

> When I use Environment Linux (BASH) : It returns error pdflatex not found

Probably the exeution process created by the _Shell commands_ plugin does not know the path to `pdflatex`. Can you please try these steps:
1. Open up a normal terminal and execute `which pdflatex`. This should give you an absolute path to `pdflatex` from the beginning of the filesystem. If not, please let me know.
2. Use the [_Add directories to the PATH environment variable_ setting](https://publish.obsidian.md/shellcommands/Environments/Additions+to+the+PATH+environment+variable#An+easier+way+to+add+directories+to+PATH) to introduce the path to SC.

Does it work now? 🙂

**Comment by tttienthinh** [comment]
Yes I have tested it : ```which pdflatex``` gives me ```/usr/bin/pdflatex```   
Inside the **Linux PATH additions** I have added ```/home/tttienthinh/.local/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/usr/games:/usr/local/games:/snap/bin``` and I have also added ```[...]:/snap/bin:/usr/bin/pdflatex```  
But always the same problem.   
I have added the switch case "sh" to ```obsidian-shellcommands/main.js``` it is now returning the same error than with bash.  

  
Thank you for you quick response ! Hope you will be able to continue to help me. :slightly_smiling_face:    
And thank you for all your work.

**Reply by tttienthinh** [reply]
I have tested something : ```which pandoc``` it gives me ```/app/bin/pandoc```  
But when I do it in the terminal, it gives me ```/usr/bin/pandoc```  
also, in the terminal, the folder ```/app``` doesn't exist.  

Hope you can help me !

**Reply by Taitava** [reply]
Strange that the problem still occurs.

> Yes I have tested it : `which pdflatex` gives me `/usr/bin/pdflatex`

Where did you try this command, in a regular terminal application or by using the _Shell commands_ plugin? I just want to ensure you tested it outside of the Obsidian application, in a normal terminal. I don't know why, but shell commands that are executed in Obsidian, do not have the same `$PATH` settings as in a regular terminal. That's why a regular terminal is needed to help us get correct paths, that can then be added to the SC plugin's settings.

## Another possible solution that came to my mind

If you have a `.bashrc` file in your home directory, you could try to add the following line in front of your shell command:
```
source ~/.bashrc
```

So your complete shell command would be:

```
source ~/.bashrc
pandoc {{file_path:absolute}} -o {{folder_path:absolute}}/{{title}}.pdf
```

This should load all possible `$PATH` additions you might have in your `.bashrc` file. Does it help?

----

> I have added the switch case "sh" to `obsidian-shellcommands/main.js` it is now returning the same error than with bash.

Yea, that's exactly what I expected. Even if that particular bug is fixed, it's not enough to fix the `$PATH` problem. But I hope I'll be able to help you solve this. 👍

**Reply by tttienthinh** [reply]
# 1
Having this switch case, can I Use system default (sh) or do I have to switch back to bash ?  
# 2 
Sorry I wasn't clear enough.  
In my terminal :   
 - ```which pandoc``` gives me ```/usr/bin/pandoc```  
 - ```which pdflatex``` gives me ```/usr/bin/pdflatex```  
  
In the plugin : 
 - ```which pandoc``` gives me ```/app/bin/pandoc```
 - ```which pdflatex``` gives me ```Error : no pdflatex in /app/bin:/usr/bin...```

# 3 
Yes I have ```.bashrc``` in my home directory. Adding it :  
```
source ~/.bashrc
which pdflatex
```
gives me ```Error : no pdflatex in /app/bin:/usr/bin...```  
AND :
```
source ~/.bashrc
pandoc {{file_path:absolute}} -o {{folder_path:absolute}}/{{title}}.pdf
```
gives me ```pdflatex not found : ... ```  
  
  
One more time, thank you very much for your help :smile: . I believe Obsidian-Shellcommands is a very good plugin for every UNIX user. Hope it will work well.

**Reply by Taitava** [reply]
> # 1
> 
> Having this switch case, can I Use system default (sh) or do I have to switch back to bash ?

Yes, you can use system default. I assume your `/bin/sh` is Bash, but even if it's something else (e.g. Zsh or Dash), this should not matter. Well, if it's something else than Bash, then `source ~/bashrc` won't work as that file is only for Bash, but now it seems the `source ~/bashrc` command did not help anyway, so it doesn't matter now.

The following does not fix anything, but if you want, you can execute this using the SC plugin and having _System default (sh)_ defined as your shell in the plugin's settings:

```
echo "Your shell is: "$(readlink /proc/$$/exe)
```

This might reveal what shell the `sh` actually is for you. As far as I know, when a default shell in a system is defined as `sh` (instead of e.g. `bash`, `dash` or `zsh`), `sh` can point to whatever unix shell. E.g. on one machine it could point to Bash, on another to Dash etc.

---

I'll continue answering to the other parts of your message later today. 🙂

**Reply by Taitava** [reply]
> # 2
> 
> Sorry I wasn't clear enough. In my terminal :
> 
>     * `which pandoc` gives me `/usr/bin/pandoc`
> 
>     * `which pdflatex` gives me `/usr/bin/pdflatex`
> 
> 
> In the plugin :
> 
>     * `which pandoc` gives me `/app/bin/pandoc`
> 
>     * `which pdflatex` gives me `Error : no pdflatex in /app/bin:/usr/bin...`
> 
> 

Thank you. Now I got this kind of idea, but I'm not sure if it will work:

```bash
alias pdflatex=/usr/bin/pdflatex
pandoc {{file_path:absolute}} -o {{folder_path:absolute}}/{{title}}.pdf
```

If you try to execute this command using the SC plugin, does it still give the `pdflatex not found : ... ` error message?

> # 3
> 
> Yes I have `.bashrc` in my home directory. Adding it :
> 
> ```
> source ~/.bashrc
> which pdflatex
> ```
> 
> gives me `Error : no pdflatex in /app/bin:/usr/bin...` AND :
> 
> ```
> source ~/.bashrc
> pandoc {{file_path:absolute}} -o {{folder_path:absolute}}/{{title}}.pdf
> ```
> 
> gives me `pdflatex not found : ... `

Ok, I guess the `/~bashrc` file does not contain any `$PATH` configs that would relate to `pdflatex`. It's completely fine, I was just hoping that including that file would solve the problem, but it didn't. We need to figure out something else.

Now that I don't have so many ideas left, I think could you make some screenshots:
 - Screenshot of the `$PATH` additions settings.
 - Screenshot of your shell command setting.
 - Screenshot of any error messages you get.
 - You can crop the images / blur some details if there's any private information you don't want to post here.

Maybe there's just some tiny bits of information that I'm not getting here.

> 
> One more time, thank you very much for your help 😄 . I believe Obsidian-Shellcommands is a very good plugin for every UNIX user. Hope it will work well.

Thank you for your feedback! ❤️ You are welcome, I'm glad you like this plugin 🙂. These "command not found" type of problems unfortunately are one of the plugin's current pain points. I'd be so glad if I could find some solution to make the plugin be able to easily access the same commands that are accessible via a normal terminal. Maybe I'll figure it out one day. 👍

**Reply by tttienthinh** [reply]
Just I don't know if you noticed it : which pandoc gives me APP/bin/pandoc,

**Reply by tttienthinh** [reply]
Not USR/bin/pandoc

**Reply by Taitava** [reply]
> Just I don't know if you noticed it : which pandoc gives me APP/bin/pandoc,
> Not USR/bin/pandoc

I didn't notice it. What happens if you run the following two commands in a _normal terminal_:
```bash
/app/bin/pandoc {{file_path:absolute}} -o {{folder_path:absolute}}/{{title}}.pdf
/usr/bin/pandoc {{file_path:absolute}} -o {{folder_path:absolute}}/{{title}}.pdf
```
(You need to replace the `{{file_path:absolute}}`, `{{folder_path:absolute}}` and `{{title}}` variables manually with the correct paths before executing the commands.)

Do both of them give the same result? Or does `/app/bin/pandoc` give a different result?

----

Also, I'd like to check if you happen to be running a different shell in Obsidian than what you have in your normal terminal. Can you execute this command:

```bash
echo "Your shell is: "$(readlink /proc/$$/exe)
```
1. Execute this using the SC plugin and having System default (sh) defined as your shell in the plugin's settings.
2. Also execute the same command in a normal terminal, so we'll see if the shells are different for some reason.

**Reply by tttienthinh** [reply]
# 1 SC plugin and having System default (sh)
```/app/bin/pandoc test.md -o test.pdf``` -> ```pdflatex not found```  
```/usr/bin/pandoc test.md -o test.pdf``` -> ```\bin\sh: /usr/bin/pandoc: No such file or directory```  
```echo "Your shell is: "$(readlink /proc/$$/exe)``` -> Nothing happen
```nautilus``` -> ```\bin\sh: nautilus: Command not found``` 


# 2 Normal Terminal
```/app/bin/pandoc test.md -o test.pdf``` -> ```bash: /app/bin/pandoc: No such file or directory```  
```/usr/bin/pandoc test.md -o test.pdf``` -> It works well  
```echo "Your shell is: "$(readlink /proc/$$/exe)``` -> ```Your shell is: /usr/bin/bash```  
```nautilus``` -> It opens my Files Explorer in Home


Thank you very much for your help, I am trying to be as clear as possible !

**Reply by tttienthinh** [reply]
![Obsidian-shell-Command.png](https://i.postimg.cc/SNJk1jVL/Obsidian-shell-Command.png)](https://postimg.cc/cryprxp6)

**Reply by Taitava** [reply]
Thank you and sorry for my delayed reply! I had a flu during the past week, so wasn't able to do anything on this project.

I don't have much time right now, so I'll come back later. But seems odd that the file `/usr/bin/pandoc` is not available when a command is executed via the plugin, but it does exist when executing via normal terminal. 🤔

**Reply by Taitava** [reply]
> # 1 SC plugin and having System default (sh)
> ...
> `echo "Your shell is: "$(readlink /proc/$$/exe)` -> Nothing happen

Ah, I forgot to instruct you to set the output channel to e.g. [Notification balloon](https://publish.obsidian.md/shellcommands/Output+handling/Output+channel+-+Notification+balloon). The output is ignored by default, so that's why nothing happened. Or, if you have the SC settings panel open, you can hold `Ctrl` while clicking _Execute now_ icon to make output be shown in a modal.

Let's see if we can get the shell's name this way. If it's something else than `bash`, it might explain why things work differently when commands are executed via the plugin.
